versions/schematic-3.md

```diff
@@ -0,0 +1,222 @@
+
+#### <a name="biomeContainer"></a> Biome Container 
+
+An object which provides biome based data and has particular requirements alone. Not all schematics need to consist of biomes.
+
+##### Fields
+
+Field Name | Type | Description
+---|:---:|---
+<a name="schematicBiomePalette"></a>Palette | [Palette Object](#paletteObject) | **Required** Specifies the palette. This is a mapping of [biomes](#defBiome) to indicies which are local to this schematic. These indices are used to reference the biomes from within the [Biome array](#schematicBiomeData). It is recommended for maximum data compression that your indices start at zero and skip no values.
+<a name="schematicBiomeData"></a>Data | `varint[]` | **Required** Specifies the main storage array which contains `Width * Height * Length` entries for `Biome`s at positions. Each entry is specified as a varint and refers to an index within [BiomePalette](#schematicBiomePalette). The entries are indexed by `x + z * Width + y * Width * Length`.
+
+#### Block State Ids
+
+The format of the Block State identifier is the id of the block type and a set of comma-separated property `key=value` pairs surrounded by square brackets. If the block has no properties then they can be excluded. The block type id is specified as a [Resource Location](#defResourceLocation).
+
+For example the air block has no properties so its id representation would be just the block type id `minecraft:air`. The planks block however has an enum property for the `variant` so its id would be `minecraft:planks[variant=oak]`. Properties should be ordered with their keys in alphabetical ordering.
+
+##### Fields
+
+Field Pattern | Type | Description
+---|:---:|---
+<a name="paletteEntry"></a>{blockstate} | `integer` | A single entry mapping a blockstate to an index.
+
+#### <a name="blockEntityObject"></a>BlockEntity Object
+
+An object to specify a `BlockEntity` which is within the region. Block entities are used by Minecraft to store additional data for a block (such as the lines of text on a sign). The fields used to describe a `BlockEntity` vary for each type, however the structure will be the same as used by the [Minecraft Chunk Format](http://minecraft.gamepedia.com/Chunk_format#Block_entity_format).
+
+##### Fields
+
+Field Pattern | Type | Description
+---|:---:|---
+<a name="blockEntityPos"></a>Pos | `integer[3]` | **Required.** The position of the `BlockEntity` relative to the `[0, 0, 0]` position of the schematic (without the [offset](#schematicOffset) applied). Must contain exactly 3 integer values.
+<a name="blockEntityId"></a>Id | `string` | **Required.** The id of the `BlockEntity` type defined by this `BlockEntity` Object, specified as a [Resource Location](#defResourceLocation). This should be used to identify which fields should be required for the definition of this type.
```

Comment:
Consider moving to `x, y, z, id` instead of `Pos, Id` because then the native format can be directly used, instead of needing to do a translation like the following:
```java
                        values.remove("id"); // Remove 'id' if it exists. We want 'Id'

                        // Positions are kept in NBT, we don't want that.
                        values.remove("x");
                        values.remove("y");
                        values.remove("z");

                        values.put("Id", new StringTag(block.getNbtId()));
                        values.put("Pos", new IntArrayTag(new int[] { x, y, z }));
```

Reply:
I prefer the separation of concerns over the entity deserializing itself off the compound and the position relative within the schematic.

Reply:
I'm still not sure that mapping `id` is really worth it, but I understand why the position is mapped.

Reply:
It's not mapping `id`, it's mapping to `Id` to remain consistent with the rest of the format being capitalized.